Add query pushdown documentation (#2339)
FerretDB/FerretDB: website/docs/pushdown.md

---

> sidebar_position: 1

Reviewer: @Fashander where is the best location for this doc?

---

# Query pushdown

**Query pushdown** is the method of optimizing a query by reducing the amount of data read and processed.
It saves memory space, network bandwidth, and reduces the query execution time by not prefetching
unnecessary data to the database management system.

> By default, FerretDB retrieves all data related to queried collection, and applies filters on its own, making

Reviewer: ```suggestion
Initially FerretDB retrieved all data related to queried collection, and applies filters on its own, making
```
I'm trying to make this sound like what we used to do and what we do now, instead of making it sounds like we still by default fetch all data.

it possible to implement complex logic safely and quickly.
To make this process more efficient, we minimize the amount of incoming data, by applying proper SQL filters.

:::info
You can learn more about query pushdown in our [blog post](https://blog.ferretdb.io/ferretdb-fetches-data-query-pushdown/).
:::

> ## Query pushdown supported filters

Reviewer: ```suggestion
## Supported types and operators
```

---

> sidebar_position: 10

Reviewer 2: Great job on the content! Since it's about performance optimization, I think we should have the position just below `operators`. @AlekSi can you advise on this?

:::

## Supported types and operators

The following table shows all operators and types that FerretDB pushdowns on PostgreSQL backend.
If filter uses type and operator, that's marked as pushdown-supported on this list,
FerretDB will prefetch less data, resulting with more performent query.

If your application requires better performance for specific operation,
feel free to share this with us on [FerretDB Slack](https://join.slack.com/t/ferretdb/shared_invite/zt-zqe9hj8g-ZcMG3~5Cs5u9uuOPnZB8~A)!

:::tip
As query pushdown allows developers to implement query optimizations separately from the features,
the table will be updated frequently.
:::

<!-- markdownlint-capture -->
<!-- markdownlint-disable MD033 MD051 MD001 -->
|            | Object   | Array   | Double                 | String   | Binary   | ObjectID   | Boolean   | Date   | Null   | Regex   | Integer   | Timestamp   | Long                   |
| ---------- | -------- | ------- | --------               | -------- | -------- | ---------- | --------- | ------ | ------ | ------- | --------- | ----------- | ------                 |

> | `=`        | ✖️        | ✖️       | ⚠️ <sub>[[1]](#1)</sub> | ✅       | ✖️        | ✅         | ✅        | ✅     | ✖️      | ✖️       | ✅        | ✖️           | ⚠️ <sub>[[1]](#1)</sub> |

Author: * MD033 (no-inline-html) disabled for <sub> tag
* MD051 (link-fragments) disabled, as `#1` is valid section in docusaurus (specified by `{#1}` at the bottom).
* MD001 (heading-increment) disabled to allow making smaller header for `[1]` annotation.

| `$gte`     | ✖️        | ✖️       | ✖️                      | ✖️        | ✖️        | ✖️          | ✖️         | ✖️      | ✖️      | ✖️       | ✖️         | ✖️           | ✖️                      |
| `$lt`      | ✖️        | ✖️       | ✖️                      | ✖️        | ✖️        | ✖️          | ✖️         | ✖️      | ✖️      | ✖️       | ✖️         | ✖️           | ✖️                      |
| `$lte`     | ✖️        | ✖️       | ✖️                      | ✖️        | ✖️        | ✖️          | ✖️         | ✖️      | ✖️      | ✖️       | ✖️         | ✖️           | ✖️                      |
| `$in`      | ✖️        | ✖️       | ✖️                      | ✖️        | ✖️        | ✖️          | ✖️         | ✖️      | ✖️      | ✖️       | ✖️         | ✖️           | ✖️                      |
| `$ne`      | ✖️        | ✖️       | ⚠️ <sub>[[1]](#1)</sub> | ✅       | ✖️        | ✅         | ✅        | ✅     | ✖️      | ✖️       | ✅        | ✖️           | ⚠️ <sub>[[1]](#1)</sub> |
| `$nin`     | ✖️        | ✖️       | ✖️                      | ✖️        | ✖️        | ✖️          | ✖️         | ✖️      | ✖️      | ✖️       | ✖️         | ✖️           | ✖️                      |

> ###### [1] {#1}

Author: Here I wanted to create reference from the table, so readers could easily click on the `[1]` inside the table to get to the longer explanation. I've tried to find some way to create an annotation, but it seems that docusaurus doesn't have anything like that. Also markdown is really simple language, and afaik it doesn't implement anything like that.

I've used headers as the dirty way of doing that. When we'll follow the header order, the text might be too large.

---
sidebar_position: 6
hide_table_of_contents: true
---

# Query pushdown

**Query pushdown** is the method of optimizing a query by reducing the amount of data read and processed.
It saves memory space, network bandwidth, and reduces the query execution time by not prefetching

> unnecessary data to the database management system.

Author: Good point! I was thinking to make some general definition here, but we could also start from a short introduction about FerretDB being proxy on top of databases like PostgreSQL.

Initially FerretDB retrieved all data related to queried collection, and applies filters on its own, making
it possible to implement complex logic safely and quickly.

> To make this process more efficient, we minimize the amount of incoming data, by applying proper SQL filters.

Reviewer 3: Minor comment (feel free to ignore).

Maybe we can say something about what "proper" is? E.g. how we make it "proper"? I understand that we want to mention things briefly and link a blog post, but still, I think we should bring more clarity, just keep it summarized. As a reader of this text, I wonder where the filters are applied and when?

In principle, the next sentences cover it, but it looks too "secret" to me here.

Author: How can we improve this description? Should I add some simple query example and how is it evaluated or it's too much for docs?

Reviewer: How about replacing `by applying proper SQL filters` to `by applying WHERE clause on SQL queries`? Is this more clear @rumyantseva?

:::info
You can learn more about query pushdown in our [blog post](https://blog.ferretdb.io/ferretdb-fetches-data-query-pushdown/).
:::

## Supported types and operators

The following table shows all operators and types that FerretDB pushdowns on PostgreSQL backend.
If filter uses type and operator, that's marked as pushdown-supported on this list,
FerretDB will prefetch less data, resulting with more performent query.

If your application requires better performance for specific operation,

> feel free to share this with us on [FerretDB Slack](https://join.slack.com/t/ferretdb/shared_invite/zt-zqe9hj8g-ZcMG3~5Cs5u9uuOPnZB8~A)!

Reviewer 4: ```suggestion
feel free to share this with us in our[community](/#community)!
```

We don't want to force Slack on people, and also we don't want to copy&paste that  non-permanent URL everywhere

:::tip
As query pushdown allows developers to implement query optimizations separately from the features,
the table will be updated frequently.
:::

<!-- markdownlint-capture -->
<!-- markdownlint-disable MD001 MD033 MD051 -->

> |            | Object   | Array   | Double                 | String   | Binary   | ObjectID   | Boolean   | Date   | Null   | Regex   | Integer   | Timestamp   | Long                   |

Reviewer 3: I remember, Alexey had some questions about this table formatting (it was too wide), but I guess it's solved?

Author: Yes, we removed the sidebar, you can see final results [here](https://c1f7bf7b.ferretdb-docs-dev.pages.dev/pushdown/)!

---
sidebar_position: 1
---

# Query pushdown

**Query pushdown** is the method of optimizing a query by reducing the amount of data read and processed.
It saves memory space, network bandwidth, and reduces the query execution time by not prefetching
unnecessary data to the database management system.

:::info
You can read more about query pushdown in our [blog post](https://blog.ferretdb.io/ferretdb-fetches-data-query-pushdown/)!
:::

## Why do we need query pushdown

As FerretDB mission is to become true open-source MongoDB-compatible database, it's important

> to handle all operations, comparisons, data types, and commands in the same fashion as MongoDB.

Reviewer 4: I think technical documentation (at that part at least) there should be more neutral and "boring". That sounds like an exiting blog post.

We mostly need to document three things:
1. What is pusdown.
2. What we support.
3. What are caveats.

Reviewer 4: Yeah, what you already wrote below
